### PR TITLE
test: Prevent non-compatible sed binary for scripted-diffs

### LINKED ISSUE #19815: improve scripted-diff check
#### Context: 

[Scripted diffs](https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md#scripted-diffs) are a way to automate some types of reformatting or refactoring changes. When a commit message matches a specific format (begins with `scripted-diff:`), CI will run [commit-script-check.sh](https://github.com/bitcoin/bitcoin/blob/master/test/lint/commit-script-check.sh). 

`sed` is platform dependent. One significant difference is when using the `-i` flag. On BSD (eg. macOS), you have to insert an empty string, but with GNU sed (used by Travis), this is compatible. 

example of a working command for these two platforms- 
macOS: `sed -i '' 's/a oneshot/an addrfetch/g' src/chainparams.cpp`
Travis: `sed -i 's/a oneshot/an addrfetch/g' src/chainparams.cpp`

#### Task: 

Add logic to the `commit-script-check.sh` to detect if the scripted diff is using the BSD sed syntax and print a helpful error message.

#### Useful skills:

Bash

#### Want to work on this issue?

For guidance on contributing, please read [CONTRIBUTING.md](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md) before opening your pull request.

### PR DESCRIPTION
**Pull request by anouarkappitou**
This Pull request improve scripted diff by checking for non-compatible sed binary. Fixes #19815

### CODE REVIEW

**Comment by MarcoFalke** [comment]
Heh, seems brittle, but sufficiently acceptable as an approximation. 

cr ACK be5286f8a2f8647f71a51d2af07bfaa70577b3d5

File: test/lint/commit-script-check.sh [hunk 1]
```diff
@@ -17,6 +17,11 @@ if test -z "$1"; then
     exit 1
 fi
 
+if ! sed --help | grep -q 'GNU'; then
+    echo "Error: the installed sed package is not compatible, please make sure you have GNU sed installed in your system";
```
Comment (shaavan) [inline test/lint/commit-script-check.sh:21]:
```suggestion
    echo "Error: the installed sed package is not compatible. Please make sure you have GNU sed installed in your system.";
```

**Review by shaavan — COMMENTED** [review]
Concept ACK

This PR runs the `sed --help` command and checks silently (i.e., without printing results) for the Keyword "GNU" presence in it. This makes sure that GNU sed is present in the system.

I agree with the approach; however, I have some suggestions to improve the error message.

File: test/lint/commit-script-check.sh (hunk 1 above)
Comment (anouarkappitou) [reply test/lint/commit-script-check.sh]: Thank you for the suggestion @shaavan

**Comment by katesalazar** [comment]
$ if ! sed --help | grep -q 'GNU'; then echo "error"; fi
    sed: illegal option -- -
    usage: sed script [-Ealnru] [-i extension] [file ...]
            sed [-Ealnu] [-i extension] [-e script] ... [-f script_file] ... [file ...]
    error
    $ ▮

Concept ACK.

**Comment by shaavan** [comment]
@anouarkappitou 
The updated error message looks good. I would recommend squashing the commits together.

**Comment by anouarkappitou** [comment]
Thank you @shaavan for the suggestions, it's been updated.

**Review by shaavan — APPROVED** [review]
ACK 30df5c3dd4d39d9027b0341d01d3233400104893

File: test/lint/commit-script-check.sh [hunk 2]
```diff
@@ -17,6 +17,11 @@ if test -z "$1"; then
     exit 1
 fi
 
+if ! sed --help | grep -q 'GNU'; then
+    echo "Error: the installed sed package is not compatible. Please make sure you have GNU sed installed in your system.";
+    exit 1;
+fi
+
```
Comment (hebasto) [inline test/lint/commit-script-check.sh:24]: > Heh, seems brittle...

Why don't check the required functionality directly:
```suggestion
echo "test" > tmp
sed -i 's/test/ok/' tmp &> /dev/null
ret=$?
rm tmp
if [ $ret -ne 0 ]; then
  echo "Error: the installed sed package is not compatible. Please make sure you have GNU sed installed in your system."
  exit 1
fi

```
?

**Comment by dongcarl** [comment]
Perhaps a point of discussion for another place, but is it possible to do `sed '/^-BEGIN VERIFY SCRIPT-$/,/^-END VERIFY SCRIPT-$/{//!b};d'` by chaining a few different UNIX string manipulation tools? Perhaps it'd be a bit more readable too?

**Comment by katesalazar** [comment]
$ sed '/^-BEGIN VERIFY SCRIPT-$/,/^-END VERIFY SCRIPT-$/{//!b};d'
    sed: 1: "/^-BEGIN VERIFY SCRIPT- ...": unexpected EOF (pending }'s)
    $ echo $?
    1

🤔 

beats me

**Comment by dongcarl** [comment]
@katesalazar I think this is exactly what this PR is addressing, you're running BSD-`sed`, which does not have the GNU extensions required to have this sed script run successfully

File: test/lint/commit-script-check.sh (hunk 2 above)
Comment (anouarkappitou) [reply test/lint/commit-script-check.sh:24]: Hello @hebasto,

Thank you for the suggestion, I was thinking to use this approach in-fact, but I avoided it due to performance reasons, please correct me if I'm wrong, by running string substitution, will require to start PCRE engine before performing string replacement, also it has overhead of creating and removing temporary file.

Comment (hebasto) [reply test/lint/commit-script-check.sh:24]: Correct. There are performance issues with this approach. But, as almost always, performance should not be considered at first place :)
I mean, are performance drawbacks a real issue?

Comment (anouarkappitou) [reply test/lint/commit-script-check.sh:24]: I see your point, if we have enough consent regarding this approach, I'll be happy to refactor the current changes.
Thank you again @hebasto
What do you think @shaavan, @katesalazar and @MarcoFalke ?

Comment (katesalazar) [reply test/lint/commit-script-check.sh:24]: I think all of these ways to do this same thing are good.
I could also concept NAK both this and the parent issue.
IDK who are we forcing this on, besides macOS users.
But I suppose macOS users should be OK to slip in their
stack just one more trusted build.

**Comment by laanwj** [comment]
Concept ACK, I've renamed the PR because `script:` is ambigious (it can also refer to bitcoin script).

**Comment by hebasto** [comment]
> Concept ACK, I've renamed the PR because `script:` is ambigious (it can also refer to bitcoin script).

Sorry, for off-topic here, but in that case https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#creating-the-pull-request should be improved, no?

**Comment by laanwj** [comment]
> Sorry for off-topic here, but in that case https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#creating-the-pull-request should be improved, no?

Maybe or maybe I'm just wrong. I still think "test:" is more appropriate for this specific PR though.

**Review by 2310Mas — APPROVED** [review]

File: test/lint/commit-script-check.sh [hunk 3]
```diff
@@ -17,6 +17,11 @@ if test -z "$1"; then
     exit 1
 fi
 
+if ! sed --help | grep -q 'GNU'; then
```
Comment (theStack) [inline test/lint/commit-script-check.sh:20]:
```suggestion
if ! sed --help 2>&1 | grep -q 'GNU'; then
```

Could redirect stderr to stdout in order to not show any output? 

On OpenBSD's sed, there is no `--help` parameter, i.e. an error message and the usage is printed on stderr:
```
$ ./test/lint/commit-script-check.sh HEAD~100
sed: unknown option -- -
usage: sed [-aEnru] [-i[extension]] command [file ...]
       sed [-aEnru] [-e command] [-f command_file] [-i[extension]] [file ...]
Error: the installed sed package is not compatible. Please make sure you have GNU sed installed in your system.
```

**Review by theStack — COMMENTED** [review]
Concept ACK

File: test/lint/commit-script-check.sh (hunk 3 above)
Comment (anouarkappitou) [reply test/lint/commit-script-check.sh]: Hello @theStack,
Thank you for the suggestion, it was applied.

**Review by shaavan — APPROVED** [review]
reACK d8dfc403f74858b79c87f650b5d609aa60a4dc3b

Changes since my last review:

- Added `2>&1` in the command to redirect standard error to standard output.

I agree with this change for the reasons @theStack discussed in [this](https://github.com/bitcoin/bitcoin/pull/23543#discussion_r777148111) comment.

p.s.: [Here](https://www.brianstorti.com/understanding-shell-script-idiom-redirect/)'s one excellent article I found about understanding `2>&1`.

**Review by theStack — APPROVED** [review]
Tested ACK d8dfc403f74858b79c87f650b5d609aa60a4dc3b

Checked that the introduced error message appears on OpenBSD 7.0 (which doesn't have GNU sed in the base system):
```
$ ./test/lint/commit-script-check.sh HEAD~100
Error: the installed sed package is not compatible. Please make sure you have GNU sed installed in your system.
$ echo $?
1
```

Also verified on Debian bookworm/sid that the script still runs if GNU sed is available.

**Pull request merged**